[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106094
* __->__ #106087

Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer

Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1154,6 +1154,49 @@ def test_xnnpack_quantizer_obs_sharing_ops(self):
+        m = M().eval()
+        example_inputs = (torch.randn(3, 5),)
+        quantizer = XNNPACKQuantizer()
+        operator_config = get_symmetric_quantization_config(is_per_channel=True)
```
This technically returns a QuantizationConfig, not an OperatorConfig right? Maybe we should rename this variable

jerryzh168 replied:
oh OK

jerryzh168 replied:
maybe let me do a follow up renaming PR

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -710,25 +780,37 @@ def _annotate_hardtanh(
+        filter_fn: Optional[Callable[[Node], bool]] = None,
     ) -> None:
         self._annotate_input_out_obs_sharing_op(
             torch.nn.AdaptiveAvgPool2d, gm, quantization_config
```
Did you forget to pass in `filter_fn` here? (and in other places)

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -417,7 +471,10 @@ def _annotate_conv2d_bn_relu(
+        filter_fn: Optional[Callable[[Node], bool]] = None,
     ) -> None:
         fused_partitions = find_sequential_partitions(
             gm, [torch.nn.Conv2d, torch.nn.ReLU]
```
What if the user has
```
self.my_conv = torch.nn.Conv2d(...)
self.my_relu = torch.nn.ReLU(...)
```
and they do `set_module_name("my_conv")` and `set_module_name("my_relu")`? I guess we ignore these settings? I guess this may be a bit complicated to support

jerryzh168 replied:
it's not used in this way, user should set_module_name("module_name_for_parent_of_conv_and_relu")

if you are talking about "quantize all conv-relu patterns in my model" I think that is another api, right now it is set_operator_type("conv_relu"), but the quantizer can expose this in a different way as well

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -263,60 +279,95 @@ def set_global(self, quantization_config: QuantizationConfig) -> XNNPACKQuantize
+        if config is None:
+            return model
+
+        assert config is not None
```
I feel we don't need this assert given the previous lines :)

jerryzh168 replied:
this is for lint I think

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -224,13 +224,29 @@ def _is_annotated(nodes: List[Node]):
     return annotated
 
 
+def _get_module_name_filter(name: str):
```
can you add a quick docstring for what this does?

andrewor14 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1154,6 +1154,49 @@ def test_xnnpack_quantizer_obs_sharing_ops(self):
+            torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+            torch.ops.aten.addmm.default,
+            torch.ops.quantized_decomposed.quantize_per_tensor.default,
+            torch.ops.quantized_decomposed.dequantize_per_tensor.default,
```
Maybe I'm missing something but I'm not sure if I understand what this test is doing. We call `set_module_name("sub", per_channel_config)`, so I would expect everything inside `sub` to be quantized per channel, but this line is saying the second linear is quantized per tensor? Why are there no `quantize_per_channels` ops? Also can you explain why is the first linear not quantized?

jerryzh168 replied:
per_channel_config is referring to the weight actually, activation is always quantized per tensor, we probably want to rename that flag to is_per_channel_weight

the first linear is not quantized because the global quantization config is None, and no quantization_config is set for that linear.

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][quantizer] Suppoert set_module_name in XNNPACKQuantizer"


Summary:
Added support to allow users to set configurations based on module name in XNNPACKQuantizer, can also serve as an example
for implementing new quantizers

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_xnnpack_quantizer_set_module_name

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
